Fix drop of pinned types

WorldSEnder opened the pull request:
A pinned value can not be moved before its drop impl is called. This is strict, so we can not copy the storage in the Drop impl. Further, the storage is now marked `PhantomPinned`, so that the compiler doesn't move it for us by accident.

A few small stylistic changes to aid readability are included too, because my linter complained too much.

In `as_pin_mut`, I'm trying hard to avoid a `transmute`, the existing API can be used, and is much more forgiving if for some reason pointer metadata or [provenance](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#pointer-provenance) is laid out differently.

I've added a test case that you can verify to complain under the previous code with miri enabled to show off the bug. A pointer to pinned data is accessed in the Drop impl, so the value must not have been moved!

WorldSEnder committed:
fix drop of pinned types

A pinned value can not be moved before its drop impl is called.
This is strict, so we can not copy the storage in the Drop impl.
Further, the storage is now marked PhantomPinned, so that the
compiler doesn't move it for us by accident.

nwtgck commented:
Great!

> I've added a test case that you can verify to complain under the previous code with miri enabled to show off the bug.

Thank you so much!

nwtgck commented on src/lib.rs:
```diff
@@ -515,5 +490,6 @@ pub mod __private_mod {
     {
         _array: ::core::mem::MaybeUninit<[u8; SIZE]>,
         _zero: <ConstUsize<ALIGN> as ToAlignedZst>::AlignedZst,
+        _pinned: ::core::marker::PhantomPinned,
```
Is it better to move `_phantom` in `__define_struct!` to here?
Then, the compiler doesn't allow `Send`, `Sync` and others for us by accident.
```rs
#[doc(hidden)]
_phantom: ::core::marker::PhantomData<(
    // for !Send + !Sync by default
    *const (),
    // for !Unpin by default
    ::core::marker::PhantomPinned,
    // for !UnwindSafe by default
    ::core::marker::PhantomData<&'static mut ()>,
    // for !Sync + !RefUnwindSafe by default
    ::core::marker::PhantomData<::core::cell::UnsafeCell<()>>,
)>,
```

WorldSEnder replied:
Good idea, I've done this!

nwtgck commented on tests/lib.rs:
```diff
@@ -465,6 +466,52 @@ async fn it_works_with_as_pin_mut() {
     assert_eq!(Future1::new().await, 10);
 }
 
+#[test]
```
I added this test without this PR fixes. Unfortunately, miri works without error. Some miri flags needed?
```
$ cargo --version
cargo 1.75.0 (1d8b05cdd 2023-11-20)
$ cargo +nightly miri --version
miri 0.1.0 (2319be8e 2024-01-12)

$ cargo +nightly miri test
...(miri outputs)...
$ echo $?
0
```

WorldSEnder replied:
I suppose you only reverted the change in the drop impl, but left the `PhantomPinned` in storage in tact.
I've added an additional test that asserts the pointer is still what we excpect it to be. Previously, with the old drop impl, the moved object contained a pointer into the not-yet-dead-but-not-moved ErasedStorage. I'm a bit suprised this passes miri with just the addition of `PhantomPinned`, but it never-the-less would violate the [pinning contract] as written.

With the lib code fully before the PR, miri would issue an error like;

```txt
test it_works_with_as_pin_drop ... error: Undefined Behavior: trying to retag from <170991> for SharedReadOnly permission at alloc46967[0x0], but that tag does not exist in the borrow stack for this location
   --> tests/lib.rs:481:34
    |
481 |                 let _ = unsafe { &*this.ptr }.to_string();
    |                                  ^^^^^^^^^^
    |                                  |
    |                                  trying to retag from <170991> for SharedReadOnly permission at alloc46967[0x0], but that tag does not exist in the borrow stack for this location
    |                                  this error occurs as part of retag at alloc46967[0x0..0x18]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <170991> was created by a SharedReadOnly retag at offsets [0x0..0x18]
   --> tests/lib.rs:490:28
    |
490 |             let self_ptr = &self.data as *const String;
    |                            ^^^^^^^^^^
help: <170991> was later invalidated at offsets [0x0..0x20] by a Unique retag
   --> tests/lib.rs:513:1
    |
513 | }
    | ^
    = note: BACKTRACE (of the first span):
    = note: inside `<it_works_with_as_pin_drop::Trap as std::ops::Drop>::drop` at tests/lib.rs:481:34: 481:44
```
```
$ cargo +nightly --version && cargo +nightly miri --version
cargo 1.77.0-nightly (7bb7b5395 2024-01-20)
miri 0.1.0 (6351247 2024-01-27)
```

[pinning contract]: https://doc.rust-lang.org/core/pin/struct.Pin.html#method.new_unchecked

WorldSEnder replied:
Note, too, that the `Trap` type is very similar to [an example in the async book](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html#pinning-to-the-stack), which should be a reasonable argument that the contained unsafes in there ought to behave correctly ^^

nwtgck replied:
Thanks!

The test is added without the fixes like the following:
https://github.com/nwtgck/stacklover-rust/compare/WorldSEnder-develop-only-test?expand=1
* commit: https://github.com/nwtgck/stacklover-rust/commit/7248a8573aabcc792c936022d2593b8816f01a2a (the branch will be removed.)
* CI result: https://github.com/nwtgck/stacklover-rust/actions/runs/7798315536/job/21266706381

The CI runs `cargo miri test`, but it did not fail.

Regardless of the results of the miri test, I would like to merge this PR.

WorldSEnder replied:
Yeah I guess miri is a bit finicky to get to consistently fail. The pointer equality check seems to trigger more reliably.

nwtgck replied:
I see.

nwtgck commented on tests/lib.rs:
```diff
@@ -465,6 +466,52 @@ async fn it_works_with_as_pin_mut() {
+        impls = (), // NOTE: !Unpin
+    }
+    let mut pinned = PinUpType::new();
+    let pinned = ::core::pin::pin!(pinned);
```
> ::core::pin::pin!

Could we use `futures::pin_mut!` instead of `core::pin::pin!`? 
Because the CI will fail because MSRV in the CI is 1.63.0.
https://github.com/nwtgck/stacklover-rust/blob/385eaa3be5b97d27451b8cc188c0d943176b9d8c/.github/workflows/ci.yml#L32-L36

Update MSRV if there is no alternative.

WorldSEnder committed:
make the test case more strict

the pointer equality fails if the data has been moved.
the read is flagged by miri if the storage is not marked as PhantomPinned

WorldSEnder committed:
move phantom tag to ErasedStorage entirely

WorldSEnder committed:
use futures::pin_mut! over pin!

std API was stabilized later, this supports more rust versions

nwtgck commented on tests/lib.rs:
```diff
@@ -465,6 +466,55 @@ async fn it_works_with_as_pin_mut() {
+    }
+    impl Drop for Trap {
+        fn drop(&mut self) {
+            let this = unsafe { Pin::new_unchecked(self) };
```
> let this = unsafe { Pin::new_unchecked(self) };

I have a question. I'm not sure why `let this = unsafe {  ... };` exists.  I assume we can use `let this = self;` instead.

WorldSEnder replied:
Technically it's okay to not do this explicitly. It serves [as a reminder that self was pinned](https://doc.rust-lang.org/std/pin/index.html#drop-implementation). Even if self was not pinned before, it doesn't hurt to assert that it will not move for the rest of the drop impl, since pinning guarantees end after drop has been called anyway.

nwtgck replied:
Thanks. Now I understood the intention.

nwtgck reviewed (approved):
Thank you for the safety improvements and detailed explanations!
